lib/internal/streams/async_iterator.js

```diff
@@ -0,0 +1,100 @@
+'use strict';
+
+const { promisify } = require('util');
+
+class Item {
+  constructor(value, done) {
+    this.done = done;
+    this.value = value;
+  }
+}
+
+function readAndResolve(iter) {
+  if (iter.lastResolve !== null) {
+    const data = iter.stream.read();
+    // we defer if data is null
+    // we can be expecting either 'end' or
+    // 'error'
+    if (data !== null) {
+      iter.lastResolve(new Item(data, false));
+      iter.lastResolve = null;
+      iter.lastReject = null;
+    }
+  }
+}
+
+class ReadableAsyncIterator {
+  constructor(stream) {
+    this.stream = stream;
+    this.lastResolve = null;
+    this.lastReject = null;
+    this.error = null;
+    this.ended = false;
+
+    stream.on('readable', () => {
+      // we wait for the next tick, because it might
+      // emit an error with process.nextTick
+      process.nextTick(readAndResolve, this);
+    });
+
+    stream.on('end', () => {
+      if (this.lastResolve !== null) {
+        this.lastResolve(new Item(null, true));
+        this.lastReject = null;
+        this.lastResolve = null;
+      }
+      this.ended = true;
+    });
+
+    stream.on('error', (err) => {
+      // reject if we are waiting for data in the Promise
+      // returned by next() and store the error
+      if (this.lastReject !== null) {
+        this.lastReject(err);
+        this.lastReject = null;
+        this.lastResolve = null;
+      }
+      this.error = err;
+    });
+
+    // the function passed to new Promise
+    // is cached so we avoid allocating a new
+    // closure at every run
+    this._handlePromise = (resolve, reject) => {
+      const data = this.stream.read();
+      if (data) {
+        resolve(new Item(data, false));
+      } else if (this.lastResolve !== null) {
+        throw new Error('next can be called only once');
```

Comment:
Hmmm... given that `readable` is handled on `nextTick`, it seems like it could be possible to have `this.lastResolve !== null` and at the same time have `data`. Maybe that condition should go first, before even calling `read()`? (And yes, I know, obscure edge case...)